src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs

```diff
@@ -7,187 +7,209 @@
-using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Elfie.Model;
+using Microsoft.CodeAnalysis.Remote;
 using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.SourceGeneration;
 using Microsoft.VisualStudio.Language.Intellisense;
 using Microsoft.VisualStudio.Shell;
 using Roslyn.Utilities;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer;
+
+internal abstract partial class BaseDiagnosticAndGeneratorItemSource : IAttachedCollectionSource
 {
-    internal abstract partial class BaseDiagnosticAndGeneratorItemSource : IAttachedCollectionSource
+    private static readonly DiagnosticDescriptorComparer s_comparer = new();
+
+    private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+    private readonly IAsynchronousOperationListener _listener;
+    private readonly BulkObservableCollection<BaseItem> _items = new();
+
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+    private readonly AsyncBatchingWorkQueue _workQueue;
+    private readonly IThreadingContext _threadingContext;
+
+    protected Workspace Workspace { get; }
+    protected ProjectId ProjectId { get; }
+    protected IAnalyzersCommandHandler CommandHandler { get; }
+
+    /// <summary>
+    /// The analyzer reference that has been found. Once it's been assigned a non-null value, it'll never be assigned
+    /// <see langword="null"/> again.
+    /// </summary>
+    private AnalyzerReference? _analyzerReference_DoNotAccessDirectly;
+
+    public BaseDiagnosticAndGeneratorItemSource(
+        IThreadingContext threadingContext,
+        Workspace workspace,
+        ProjectId projectId,
+        IAnalyzersCommandHandler commandHandler,
+        IDiagnosticAnalyzerService diagnosticAnalyzerService,
+        IAsynchronousOperationListenerProvider listenerProvider)
     {
-        private static readonly DiagnosticDescriptorComparer s_comparer = new DiagnosticDescriptorComparer();
+        _threadingContext = threadingContext;
+        Workspace = workspace;
+        ProjectId = projectId;
+        CommandHandler = commandHandler;
+        _diagnosticAnalyzerService = diagnosticAnalyzerService;
+
+        _listener = listenerProvider.GetListener(FeatureAttribute.SourceGenerators);
+        _workQueue = new AsyncBatchingWorkQueue(
+            DelayTimeSpan.Idle,
+            ProcessQueueAsync,
+            _listener,
+            _cancellationTokenSource.Token);
+    }
 
-        private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+    protected AnalyzerReference? AnalyzerReference
+    {
+        get => _analyzerReference_DoNotAccessDirectly;
+        set
+        {
+            Contract.ThrowIfTrue(_analyzerReference_DoNotAccessDirectly != null);
```

Comment:
we now control this entrypoint as the single way that a subclass can set this.  We still preserve teh semantics that it is only set once.

Reply:
see comment at https://github.com/dotnet/roslyn/pull/74444/files#diff-1341f3b2de264a2915d858c70421e1ed7c062bb17a3bd541ac466811b8c861e9L21 for why this is ok.